File: bin/python3/Predict the Winner.py
class Solution:
    def PredictTheWinner(self, nums: List[int]) -> bool:
        def play(nums, left, right):
            if left > right:
                return 0
            
            score_left = nums[left] - play(nums, left+1, right)
            score_right = nums[right] - play(nums, left, right-1)
            
            return max(score_left, score_right)
        
        return play(nums, 0, len(nums) - 1) >= 0

File: bin/go/Predict the Winner.go
package main

func PredictTheWinner(nums []int) bool {
    return play(nums, 0, len(nums)-1) >= 0
}

func play(nums []int, left, right int) int {
    if left > right {
        return 0
    }

    scoreLeft := nums[left] - play(nums, left+1, right)
    scoreRight := nums[right] - play(nums, left, right-1)

    if scoreLeft > scoreRight {
        return scoreLeft
    }
    return scoreRight
}